src/main/kotlin/org/rust/ide/annotator/RsErrorAnnotator.kt

```diff
@@ -884,6 +884,14 @@ class RsErrorAnnotator : AnnotatorBase(), HighlightRangeExtension {
         RsDiagnostic.LoopOnlyKeywordUsedOutsideOfLoopError(expr.operator).addToHolder(holder)
     }
 
+    private fun checkArrayType(holder: RsAnnotationHolder, o: RsArrayType) {
+        collectDiagnostics(holder, o)
+        val expr = o.expr
+        if (expr != null && o.arraySize == null) {
```

Comment:
We can't rely on `arraySize` here. If `arraySize` is null it doesn't exactly mean that a user uses non-constant expression because,  unfortunately, our constant evaluator may fail because of other reasons.
For example, you can call const function as expression but the plugin can't calculate it for now - #4738.
Or, it may fail because of bug (see #4738).
So, we have to check it manually - walk through array size expression and check that whole expression can be evaluated in compile time. i.e. only constants, literals, const functions, etc. are used

Reply:
Yes, but also need to override `visitArrayExpr` to check array expr.